.github/workflows/ci.yml

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) 2023 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+name: CI
+on:
+  # See: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request.
+  pull_request:
+  # See: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#push.
+  push:
+    branches:
+      - '**'
+    tags-ignore:
+      - '**'
+
+env:
+  DANGER_RUN_CI_ON_HOST: 1
+  TEST_RUNNER_TIMEOUT_FACTOR: 40
+
+jobs:
+  macos-cross:
+    name: "macOS 11.0 [gui, no tests] [jammy]"
+    runs-on: ubuntu-latest
+
+    # No need to run on the read-only mirror, unless it is a PR.
+    if: github.repository != 'bitcoin-core/gui' || github.event_name == 'pull_request'
+
+    timeout-minutes: 120
+
+    env:
+      MAKEJOBS: '-j4'
+      CI_IMAGE_NAME_TAG: 'ubuntu:jammy'
+      FILE_ENV: './ci/test/00_setup_env_mac.sh'
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+        with:
+          # See: https://github.com/moby/buildkit/issues/3969.
+          driver-opts: |
+            network=host
+
+      - name: Build container
+        uses: docker/build-push-action@v4
+        with:
+          context: .
+          file: ./ci/test_imagefile
+          build-args: |
+            CI_IMAGE_NAME_TAG
+            FILE_ENV
+          tags: test_image
+          load: true
+          cache-from: type=gha
+          cache-to: type=gha,mode=min
```

Comment:
https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy:

> ... the total size of all caches in a repository is limited to 10 GB.

Reply:
> So every image will be re-built every week regardless of whether it changed?

Sorry, I don't follow. Why will it be re-built every week?

Reply:
> New question: So every image will be re-built regardless of whether it changed...

If no changes were introduced, the image is built from the cached layers.

Reply:
Yes, it will happen from time to time. But not every run :)

There is an alternative to that approach, which is using Container Registry for images.

Reply:
> What about artefacts? (Ref. discussion in the [bitcoin-core/secp256k1#1398 (comment)](https://github.com/bitcoin-core/secp256k1/pull/1398#issuecomment-1677074743))

Artifacts are subjects of 90 days [retention period](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts). Therefore, I still lean to ghct.io.

Reply:
> Also, is the 10 GB enough to store all ccache + depends + image + ... stuff for all tasks?

Well, it depends on what "all tasks" encompass.

Anyway, ccache caches are to be updated on every run (that is not the case in libsecp repo, btw), which means that the current approach to handle Docker images is suboptimal (

Reply:
Yes, artifacts will be deleted after 90 days, but the problem with artifacts is that [they can only be downloaded by jobs in the same *run*](https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts#downloading-or-deleting-artifacts). So artifacts are a good way to pass data around within the same run, but they cannot be used for caching across runs.

But I agree that using GHCR feels wrong.... Its purpose is to publish images, so it has a bit of an "official" character. The packages will listed on the repo page for example: https://github.com/hebasto/secp256k1/ (@hebasto Can this be disabled?) On the other hand, it's possible to delete images.

Reply:
You, mean also for the Docker build cache?  Yeah, that's a damn simple approach. Then 10 GB are enough for sure, and okay, PRs that modify the Dockerfile will rebuild the Docker image on every push time, but seems acceptable. Anyway, if working on the images, it makes sense to test/build them locally first. 

But wouldn't you need artifacts again to transfer the image from the Docker build task to the actual CI task, namely in PRs that modify the Dockerfile, to avoid building the image in every single CI task? Or is this acceptable for this repo, because there won't be many tasks? 

The only theoretical problem is this:
 > GitHub will remove any cache entries that have not been accessed in over 7 days. 

It's rather unlikely in this repo that there will be no PR for 7 days. ;) It may happen in the secp256k1 repo, then we could simply have a cron job that runs once every week to rebuild the Docker image. (This is anyway not a bad idea if we want to rebuild our compiler snapshots.)